src/createActionListenerMiddleware.ts

```diff
@@ -0,0 +1,332 @@
+  <
+    C extends TypedActionCreator<any>,
+    S,
+    D extends Dispatch,
+    O extends ActionListenerOptions
+  >(
+    actionCreator: C,
+    listener: ActionListener<ReturnType<C>, S, D, O>,
+    options?: O
+  ): AddListenerAction<ReturnType<C>, S, D, O>
+
+  <S, D extends Dispatch, O extends ActionListenerOptions>(
+    type: string,
+    listener: ActionListener<AnyAction, S, D, O>,
+    options?: O
+  ): AddListenerAction<AnyAction, S, D, O>
+}
+
+interface RemoveListenerAction<
+  A extends AnyAction,
+  S,
+  D extends Dispatch<AnyAction>
+> {
+  type: 'actionListenerMiddleware/remove'
+  payload: {
+    type: string
+    listener: ActionListener<A, S, D, any>
+  }
+}
+
+/**
+ * @alpha
+ */
+export const removeListenerAction = createAction(
+  'actionListenerMiddleware/remove',
+  function prepare(
+    typeOrActionCreator: string | TypedActionCreator<string>,
+    listener: ActionListener<any, any, any, any>
+  ) {
+    const type =
+      typeof typeOrActionCreator === 'string'
+        ? typeOrActionCreator
+        : (typeOrActionCreator as TypedActionCreator<string>).type
+
+    return {
+      payload: {
+        type,
+        listener
+      }
+    }
+  }
+) as BaseActionCreator<
+  { type: string; listener: ActionListener<any, any, any, any> },
+  'actionListenerMiddleware/remove'
+> & {
+  <C extends TypedActionCreator<any>, S, D extends Dispatch>(
+    actionCreator: C,
+    listener: ActionListener<ReturnType<C>, S, D, any>
+  ): RemoveListenerAction<ReturnType<C>, S, D>
+
+  <S, D extends Dispatch>(
+    type: string,
+    listener: ActionListener<AnyAction, S, D, any>
+  ): RemoveListenerAction<AnyAction, S, D>
+}
+
+/**
+ * @alpha
+ */
+export function createActionListenerMiddleware<
+  S,
+  D extends Dispatch<AnyAction> = Dispatch
+>() {
+  type ListenerEntry = ActionListenerOptions & {
+    listener: ActionListener<any, S, D, any>
+  }
+
+  const listenerMap: Record<string, Set<ListenerEntry> | undefined> = {}
+  const middleware: Middleware<
+    {
+      (action: Action<'actionListenerMiddleware/add'>): Unsubscribe
+    },
+    S,
+    D
+  > = api => next => action => {
+    if (addListenerAction.match(action)) {
+      const unsubscribe = addListener(
+        action.payload.type,
+        action.payload.listener,
+        action.payload.options
+      )
+
+      return unsubscribe
+    }
+    if (removeListenerAction.match(action)) {
+      removeListener(action.payload.type, action.payload.listener)
+
+      return
+    }
+
+    const listeners = listenerMap[action.type]
+    if (listeners) {
+      const defaultWhen = 'after'
+      let result: unknown
+      for (const phase of ['before', 'after'] as const) {
+        for (const entry of listeners) {
+          if (phase !== (entry.when || defaultWhen)) {
+            continue
+          }
+          let stoppedPropagation = false
+          let currentPhase = phase
+          let synchronousListenerFinished = false
+          entry.listener(action, {
```

Comment:
Sorry, I'm a bit late :)

### Issue

Have you considered what happens if a listener throws an exception?

In the current implementation an exception raised in a listener would act like a [stopImmediatePropagation](https://developer.mozilla.org/en-US/docs/Web/API/Event/stopImmediatePropagation).

I do not think that is desirable to to have this behaviour especially when an exception is thrown unintentionally (e.g. `cannot read property X of undefined`).

In big apps it is extremely confusing and frustrating when an unrelated piece of code causes bugs to a feature.

### Example

```ts
listeners = [
  trackUserClick, // analytics e.g. call gtag or something similar...
  respondToUserClick // business logic
];
```

1. User  interaction triggers an event.
2. trackUserClick raises an exception for some reason:  
    integration error server down, `cannot read property X of undefined`, bug etc... 
3. respondToUserClick does not run, we have failed to respond to user interaction due to **implicit** `stopImmediatePropagation` behaviour.

---

### Other event listener APIs

#### DOM addEventlistener

Each listener runs in its own task, a subscriber exception does not stop the propagation.

#### Svelte stores

[It does not handle  subscriber exceptions](https://github.com/sveltejs/svelte/blob/master/src/runtime/store/index.ts#L73), unfortunately.

#### Observer proposal

[A subscriber cannot stop the propagation](https://github.com/tc39/proposal-observable/blob/master/src/Observable.js#L49)

#### rxjs

[A subscriber cannot stop the propagation](https://github.com/ReactiveX/rxjs/blob/6fafcf53dc9e557439b25debaeadfd224b245a66/src/internal/Subscriber.ts#L177) errors are reported in [another task](https://github.com/ReactiveX/rxjs/blob/6fafcf53dc9e557439b25debaeadfd224b245a66/src/internal/util/reportUnhandledError.ts#L13)


#### Suggested changes

I believe this event loop should call every listener in a try catch and should then notify the errors.

A global report handler should suffice, consider [redux-saga onError as an example](https://redux-saga.js.org/docs/api#createsagamiddlewareoptions),
but every listener could also  have the option to provide an error handler. 

The latter is implemented in this [event listener middleware I wrote few day ago](https://gist.github.com/FaberVitale/476e2977f59c38a703aa63cca56beaf7).